frameworklauncher/src/main/java/com/microsoft/frameworklauncher/common/utils/ValueRangeUtils.java

```diff
@@ -386,16 +390,43 @@ public static String toEnviromentVariableString(List<ValueRange> portRanges, Map
           portsString.append(";");
         } else {
           //if user not specified ports, assign the allocated ContainerPorts to each port label.
-          List<ValueRange> assignPorts = ValueRangeUtils.getSubRangeSequence(portRanges, ports.getCount(), basePort);
-          basePort = assignPorts.get(assignPorts.size() - 1).getEnd() + 1;
-          portsString.append(key + ":" + assignPorts.get(0).toDetailString(","));
-          for (int i = 1; i < assignPorts.size(); i++) {
-            portsString.append("," + assignPorts.get(i).toDetailString(","));
+          List<ValueRange> assignPorts = ValueRangeUtils.getSubRangeSequence(localPortRanges, ports.getCount(), 0);
+          if (getValueNumber(assignPorts) == ports.getCount()) {
+            localPortRanges = ValueRangeUtils.subtractRange(localPortRanges, assignPorts);
+
+            portsString.append(key + ":" + assignPorts.get(0).toDetailString(","));
+            for (int i = 1; i < assignPorts.size(); i++) {
+              portsString.append("," + assignPorts.get(i).toDetailString(","));
+            }
+            portsString.append(";");
+          } else {
+            throw new NotAvailableException("there is no enough ports to meet portsDefinitions requests");
           }
```

Comment:
for better debug, can we add portsDefinitions and portRanges into the exception message?